storage/postgreskv/client.go

```diff
@@ -317,3 +318,67 @@ func (client *Client) Iterate(ctx context.Context, opts storage.IterateOptions,
 
 	return fn(ctx, opi)
 }
+
+// CompareAndSwap atomically compares and swaps oldValue with newValue
+func (client *Client) CompareAndSwap(ctx context.Context, key storage.Key, oldValue, newValue storage.Value) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return client.CompareAndSwapPath(ctx, storage.Key(defaultBucket), key, oldValue, newValue)
+}
+
+// CompareAndSwapPath atomically compares and swaps oldValue with newValue in the given bucket
+func (client *Client) CompareAndSwapPath(ctx context.Context, bucket, key storage.Key, oldValue, newValue storage.Value) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	if key.IsZero() {
+		return storage.ErrEmptyKey.New("")
+	}
+
+	tx, err := client.pgConn.BeginTx(ctx, nil)
```

Comment:
This should be possible without transactions.

`UPDATE pathdata(bucket, fullpath, metadata) VALUES ..... WHERE metadata = old-value`, this will fail when there is no such old value/

Reply:
Unfortunately, it won't fail, but it will be successful with 0 rows affected.

I can check if 0 rows are affected, but I cannot determine if this is due to a missing key (where we should return `ErrKeyNotFound`), or due to changed value (where we should return `ErrValueChanged`).

Any idea how to implement this without transaction?

Reply:
@thepaul any ideas?

Reply:
The best thing I can think of here would be a CTE, looking something like

```
WITH matching_key AS (
    SELECT * FROM pathdata WHERE bucket = $1 AND fullpath = $2
), updated AS (
    UPDATE pathdata
        SET metadata = $3
        FROM matching_key mk
        WHERE pathdata.metadata = $4
            AND pathdata.bucket = mk.bucket
            AND pathdata.fullpath = mk.fullpath
        RETURNING 1
)
SELECT EXISTS(SELECT 1 FROM matching_key) AS key_present, EXISTS(SELECT 1 FROM updated) AS value_set;
```

If it returns `t, t` then the key was present and the value was updated.
If it returns `t, f` then the key was present but the value did not match expectations, so it was not updated.
If it returns `f, f` then the key was not present.
If it returns `f, t` then your database system is on fire.

Reply:
When value is nil and you want a delete instead, then it would be mostly the same, but the `updated` CTE would look like this instead:

```
    DELETE FROM pathdata
        USING matching_key mk
        WHERE pathdata.metadata = $4
            AND pathdata.bucket = mk.bucket
            AND pathdata.fullpath = mk.fullpath
        RETURNING 1
```

I see that this function also falls back on an `INSERT` sometimes, but the semantics there seem confusing to me in the context of a compare-and-swap operation, so I'm not entirely sure how to translate that.

Reply:
You could self join and use a returning clause ala https://stackoverflow.com/questions/7923237/return-pre-update-column-values-using-sql-only-postgresql-version

Reply:
^^ this starts to enter the realm of hard-to-understand code though

Reply:
@wthorp That's sort of what I'm doing above, but in a more readable way imo. Or does that approach buy us something else I'm not seeing yet?

Reply:
Mostly I just didn't refresh / see your post before I posted.  I like readable.  :D

Reply:
@thepaul 
> I see that this function also falls back on an INSERT sometimes, but the semantics there seem confusing to me in the context of a compare-and-swap operation, so I'm not entirely sure how to translate that.

This is the case of `CompareAndSwap(ctx, "some-key", nil, "some-value")`, i.e. the key is set to the value only if the key did not exist yet.

Reply:
@thepaul thanks for your suggestions! I believe I applied them correctly in https://github.com/storj/storj/pull/2602/commits/e60a89a7f72f3349adbe5214762cae0325e83ea4. All tests are fine.